### PR TITLE
Use general range reasoning in the cost estimator

### PR DESCRIPTION
**Pull request by bojanserafimov**
Context: The paginator returns a query describing the first page, and a query describing the rest of the results. The paginator can then be ran again on the remainder, producing the second page, and the remainder. This process continues until we run out of pages. However, to make sure we eventually run out of pages, the cost estimator must be able to understand any filters produced by the paginator. A useful sanity check is that the estimated number of pages of a paginator-generated page is one page.

To satisfy this property, both the paginator and the cost estimator will use the same range reasoning method, using the `int_value_conversion.py` module. In this PR I integrate this module into the cost estimator. In future PRs, the paginator will be built using the same `int_value_conversion` methods.

Since the generic range reasoning methods can reason about datetimes, the functionality of the cost estimator is implicitly extended, so I add a test with cost estimation on datetimes.

### CODE REVIEW

File: graphql_compiler/cost_estimation/filter_selectivity_utils.py [hunk 1]
```diff
@@ -395,6 +394,33 @@ def _get_selectivity_fraction_of_interval(interval, quantiles):
...
+                    parameters[get_parameter_name(filter_argument)],
+                )
+                for filter_argument in filter_info.args
+            ]
+
+            filter_interval = _get_query_interval_of_integer_inequality_filter(
+                parameter_values, filter_info.op_name
+            )
+            interval = _get_intersection_of_intervals(interval, filter_interval)
+    return interval
```
Comment (bojanserafimov) [inline graphql_compiler/cost_estimation/filter_selectivity_utils.py]: moved from line 428..

File: graphql_compiler/cost_estimation/filter_selectivity_utils.py [hunk 2]
```diff
@@ -463,22 +472,31 @@ def get_selectivity_of_filters_at_vertex(schema_info, filter_infos, parameters,
                 selectivity_at_field = _combine_filter_selectivities(
                     [selectivity_at_field, selectivity]
                 )
-            elif is_int_field:
-                # int fields are not uniformly distributed but we have quantile information.
+            else:
+                # Get value interval
+                # HACK(bojanserafimov): Abuse of IntegerInterval spec. Constructed with non-interval
+                #                       endpoints. The type should be renamed to allow this usage.
```
Comment (bojanserafimov) [inline graphql_compiler/cost_estimation/filter_selectivity_utils.py]: Will be addressed in this PR

**Comment by obi1kenobi** [comment]
Seems fine so far.

File: graphql_compiler/cost_estimation/filter_selectivity_utils.py [hunk 3]
```diff
@@ -22,14 +31,15 @@
     ),
 )
 
-# IntegerInterval is used to denote a continuous non-empty interval of integers.
-IntegerInterval = namedtuple(
-    "IntegerInterval",
+# Interval is used to denote a continuous non-empty interval of values.
```
Comment (bojanserafimov) [inline graphql_compiler/cost_estimation/filter_selectivity_utils.py]: I could make it generic, but then it wouldn't be a namedtuple. Mypy doesn't support generic namedtuples.

Comment (bojanserafimov) [reply graphql_compiler/cost_estimation/filter_selectivity_utils.py]: Doing this in a separate PR.. Looks like it's worth it

Comment (bojanserafimov) [reply graphql_compiler/cost_estimation/filter_selectivity_utils.py]: https://github.com/kensho-technologies/graphql-compiler/pull/726

Comment (obi1kenobi) [reply graphql_compiler/cost_estimation/filter_selectivity_utils.py]: Generic dataclasses are a thing, though — and I think that might be worth looking into.

**Comment by bojanserafimov** [comment]
Waiting on https://github.com/kensho-technologies/graphql-compiler/pull/726

File: graphql_compiler/cost_estimation/filter_selectivity_utils.py [hunk 4]
```diff
@@ -353,22 +362,29 @@ def get_selectivity_of_filters_at_vertex(schema_info, filter_infos, parameters,
...
+                lower_bound, upper_bound = None, None
+                if interval.lower_bound is not None:
+                    lower_bound = convert_int_to_field_value(
+                        schema_info, location_name, field_name, interval.lower_bound
+                    )
+                if interval.upper_bound is not None:
+                    upper_bound = convert_int_to_field_value(
+                        schema_info, location_name, field_name, interval.upper_bound
+                    )
+                value_interval = Interval[int](lower_bound, upper_bound)
```
Comment (bojanserafimov) [inline graphql_compiler/cost_estimation/filter_selectivity_utils.py]: This is incorrect, and shouldn't pass type checks. Fixing..

**Pull request merged**